### PR TITLE
LPS-83359 Filter provider for Structured Content

### PR DESCRIPTION
**Pull request by darquesdev**

### CODE REVIEW

**Comment by darquesdev** [comment]
cc @ruben-pulido @ahdezma @juliocamarero

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/FilterParserImpl.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.apache.olingo.server.api.uri.UriInfo;
+import org.apache.olingo.server.api.uri.queryoption.FilterOption;
+import org.apache.olingo.server.api.uri.queryoption.expression.Expression;
+import org.apache.olingo.server.core.uri.parser.Parser;
+
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * <code>FilterParser</code> transforms a sequence of characters in an manageable
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/FilterParserImpl.java]: gr formatJavadoc

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/InvalidFilterException.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * This exception is used to model errors when handling search {@link
+ * com.liferay.structured.content.apio.architect.filter.Filter}s.
+ *
+ * @author David Arques
+ * @review
+ */
+public class InvalidFilterException extends BadRequestException {
+
+	public InvalidFilterException(String s) {
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/InvalidFilterException.java]: Usually the string is called msg but s is also Ok

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/InvalidFilterException.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		super(s);
+	}
+
+	public InvalidFilterException(String message, Throwable cause) {
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/InvalidFilterException.java]: But it should be the same in both constructors, so msg :)

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	public String getSingleEntityTypeName() {
+		return _ENTITY_NAME;
+	}
+
+	private CsdlEntityContainer _createCsdlEntityContainer() {
+		CsdlEntityContainer csdlEntityContainer = new CsdlEntityContainer();
+
+		CsdlEntityType csdlEntityType = _createCsdlEntityType();
+
+		FullQualifiedName fullQualifiedName = new FullQualifiedName(
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java]: I don't know if the typical "defined when used" applies here

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+			_NAMESPACE, csdlEntityType.getName());
+
+		CsdlEntitySet csdlEntitySet = new CsdlEntitySet();
+
+		csdlEntitySet.setName(csdlEntityType.getName());
+		csdlEntitySet.setType(fullQualifiedName);
+
+		csdlEntityContainer.setName(_ENTITY_NAME);
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java]: but surely applies here, container initialisation should be here

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+			Collections.singletonList(_createCsdlEntityType()));
+
+		csdlSchema.setEntityContainer(_createCsdlEntityContainer());
+
+		return csdlSchema;
+	}
+
+	private static final String _ENTITY_NAME = "StructuredContent";
+
+	private static final String _NAMESPACE = "HeadlessApio";
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java]: does the namespace have to be in upper camel case? I've seen string constants in lower case and separated by _

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/test/java/com/liferay/structured/content/apio/internal/provider/StructuredContentEdmProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import org.junit.Test;
+
+/**
+ * @author David Arques
+ */
+public class StructuredContentEdmProviderTest {
+
+	@Test
+	public void testGetSingleEntityTypeName() {
+		StructuredContentEdmProvider provider =
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/test/java/com/liferay/structured/content/apio/internal/provider/StructuredContentEdmProviderTest.java]: I think you should use the type as the name :(

File: modules/apps/headless-apio/structured-content/structured-content-apio-api/src/main/java/com/liferay/structured/content/apio/architect/filter/Filter.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.structured.content.apio.architect.filter;
+
+import org.apache.olingo.server.api.uri.queryoption.expression.Expression;
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-api/src/main/java/com/liferay/structured/content/apio/architect/filter/Filter.java]: mmm we are coupling our API with an external library... I suppose that is the "one" that has everything we need :)

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/FilterParserImpl.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+ * OData {@link Expression}.
+ *
+ * @author David Arques
+ * @review
+ */
+@Component(immediate = true, service = FilterParser.class)
+public class FilterParserImpl implements FilterParser {
+
+	@Override
+	public Expression parse(String rawFilter) {
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/FilterParserImpl.java]: Why not filter (already used that way on the tests), or filterString (following Sort pattern)

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+	private CsdlEntityType _createCsdlEntityType() {
+		CsdlProperty titleCsdlProperty = _createCsdlProperty(
+			_TITLE_FIELD, EdmPrimitiveTypeKind.String.getFullQualifiedName());
+
+		CsdlEntityType csdlEntityType = new CsdlEntityType();
+
+		csdlEntityType.setName(_ENTITY_NAME);
+
+		List<CsdlProperty> csdlPropertyList = new ArrayList<>();
```
Comment (nhpatt) [inline modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java]: plural also works, csdlProperties

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java (hunk 6 above)
Comment (nhpatt) [reply modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/StructuredContentEdmProvider.java]: But if they are public, it's weird. Better leave it that way if they are part of the API.

File: modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/FilterParserImpl.java (hunk 9 above)
Comment (nhpatt) [reply modules/apps/headless-apio/structured-content/structured-content-apio-impl/src/main/java/com/liferay/structured/content/apio/internal/architect/filter/FilterParserImpl.java]: Not sure about this one (but definitely not rawFilter), but let's play :)

**Pull request merged**